packages/@lwc/ssr-compiler/src/transmogrify.ts

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright (c) 2024, Salesforce, Inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import { traverse, builders as b } from 'estree-toolkit';
+import { produce } from 'immer';
+import type { Node } from 'estree';
+import type { Program as EsProgram } from 'estree';
+
+export type TransmogrificationMode = 'sync' | 'async';
+
+interface TransmogrificationState {
+    mode: TransmogrificationMode;
+}
+
+export type Visitors = Parameters<typeof traverse<Node, TransmogrificationState>>[1];
+
+const EMIT_IDENT = b.identifier('$$emit');
+
+const visitors: Visitors = {
+    $: { scope: true },
+    FunctionDeclaration(path, state) {
+        const { node } = path;
+        if (!node?.async || !node?.generator) {
+            return;
+        }
+        node.generator = false;
```

Comment:
We might want to add additional guards here to ensure that non-generated code is not transformed from async-generator to something else. For example, if the customer added an async generator function to their component source, it would be transformed here.

The additional guards would probably check for function names that match either `/tmpl($\d+)?/` or `/generateMarkup($d\+)?/`.

Reply:
I addressed this with some additional compilation-time protections.